From fb0f72afe30d34b8c102cabbe87a6afdf7aad9eb Mon Sep 17 00:00:00 2001
From: Varsha Maira <vmaira@habana.ai>
Date: Thu, 13 Jun 2024 22:27:41 +0300
Subject: [PATCH] [FSW-418] fs1: bootrom: init tpm spi

This patch decouples TPM SPI init from tpm2_init().
TPM SPI initialization will skipped if TPM extend
operation was completed in previous resets and does
not need to be repeated.

Change-Id: I81c35f2a93995726f4453f29fa0f430fbb7add24
Signed-off-by: Varsha Maira <vmaira@habana.ai>
---

diff --git a/app/boot_rom.c b/app/boot_rom.c
index 3326097..b228fe4 100644
--- a/app/boot_rom.c
+++ b/app/boot_rom.c
@@ -33,6 +33,8 @@
 
 	boot_rom_fuse();
 
+	boot_rom_tpm_spi_setup();
+
 	boot_rom_security_init();
 
 	boot_rom_wait_for_signal();
diff --git a/common/common.h b/common/common.h
index 4a26e15..97e9dcc 100644
--- a/common/common.h
+++ b/common/common.h
@@ -27,5 +27,6 @@
 uint8_t boot_rom_get_tpm_next_pcr_index(void);
 void boot_rom_set_tpm_next_pcr_index(uint8_t tpm_pcr_index);
 void boot_rom_tpm_next_pcr_index_init(void);
+uint32_t boot_rom_tpm_spi_baudr_get(void);
 
 #endif /* _COMMON_H_ */
diff --git a/common/spi.c b/common/spi.c
index 97785e1..89f2ae1 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -315,28 +315,16 @@
 	spi_dw_writel(ctx, 0, mmSPI_TXFTLR);
 }
 
-#define DW_SPI_BAUD_RATE		4
-__weak uint32_t hl_spi_baudr_get(void)
-{
-	return DW_SPI_BAUD_RATE;
-}
-
 void spi_init(struct spi_ctx *ctx)
 {
-	uint32_t baudr;
-
 	if (ctx->spi_initialized)
 		return;
-
-	baudr = hl_spi_baudr_get();
-	/* Validate that baudr is even */
-	baudr = (baudr + 1) & ~1;
 
 	ctx->cs_init();
 	/* Disable ssi module */
 	spi_dw_writel(ctx, 0x00000000, mmSPI_SSIENR);
 	/* Set Baud rate */
-	spi_dw_writel(ctx, baudr, mmSPI_BAUDR);
+	spi_dw_writel(ctx, ctx->baudrate, mmSPI_BAUDR);
 	/* Set Rx FIFO threshold value to default value */
 	spi_dw_writel(ctx, 0x00000000, mmSPI_RXFTLR);
 	/* Set Tx FIFO threshold value to default value */
diff --git a/common/spi.h b/common/spi.h
index 606a655..e6bb1ba 100644
--- a/common/spi.h
+++ b/common/spi.h
@@ -17,6 +17,7 @@
 	bool spi_initialized;
 	bool spi_access_is_locked;
 	uint32_t fifo_len;
+	uint32_t baudrate;
 	bool spi_dword_access_support;
 	void (*cs_init)();
 	void (*claim_bus)();
diff --git a/common/tpm2.c b/common/tpm2.c
index 9b5c626..92495e9 100644
--- a/common/tpm2.c
+++ b/common/tpm2.c
@@ -73,14 +73,14 @@
 		tx_buf[2] = addr >> 8;
 		tx_buf[3] = addr;
 
-		ret = spi_xfer(&tpm_spi, 4 * 8, tx_buf, rx_buf, SPI_XFER_BEGIN);
+		ret = spi_xfer(boot_rom_global.tpm_spi, 4 * 8, tx_buf, rx_buf, SPI_XFER_BEGIN);
 		if (ret)
 			goto release_bus;
 
 		/* Wait state */
 		if (!(rx_buf[3] & 0x1)) {
 			for (i = 0; i < TPM_WAIT_STATES; i++) {
-				ret = spi_xfer(&tpm_spi, 1 * 8, NULL, rx_buf,
+				ret = spi_xfer(boot_rom_global.tpm_spi, 1 * 8, NULL, rx_buf,
 						0);
 				if (ret)
 					goto release_bus;
@@ -103,7 +103,7 @@
 			memset(tx_buf, 0, transfer_len);
 		}
 
-		ret = spi_xfer(&tpm_spi, transfer_len * 8,
+		ret = spi_xfer(boot_rom_global.tpm_spi, transfer_len * 8,
 			       tx_buf,
 			       in ? rx_buf : NULL,
 			       SPI_XFER_END);
@@ -153,7 +153,13 @@
 	regmap_writel(reg, out_en);
 }
 
-static void tpm_spi_init(void)
+#define DW_TPM_SPI_BAUD_RATE		4
+uint32_t boot_rom_tpm_spi_baudr_get(void) /*TODO : get baudr from fuse setting*/
+{
+	return DW_TPM_SPI_BAUD_RATE;
+}
+
+static void boot_rom_tpm_spi_local_ctx_init(void)
 {
 	tpm_spi.spi_base = spi_tpm_base_get();
 	tpm_spi.spi_initialized = false;
@@ -162,7 +168,28 @@
 	tpm_spi.claim_bus = spi_tpm_claim_bus;
 	tpm_spi.release_bus = spi_tpm_release_bus;
 	tpm_spi.fifo_len = 0;
-	spi_init(&tpm_spi);
+	tpm_spi.baudrate = boot_rom_tpm_spi_baudr_get();
+	/* Validate that baudr is even */
+	tpm_spi.baudrate = (tpm_spi.baudrate + 1) & ~1;
+}
+
+void boot_rom_tpm_spi_global_ctx_init(struct spi_ctx *tpm_spi)
+{
+	boot_rom_global.tpm_spi = tpm_spi;
+}
+
+void boot_rom_tpm_spi_init(void)
+{
+	spi_init(boot_rom_global.tpm_spi);
+}
+
+void boot_rom_tpm_spi_setup(void)
+{
+	boot_rom_tpm_spi_local_ctx_init();
+
+	boot_rom_tpm_spi_global_ctx_init(&tpm_spi);
+
+	boot_rom_tpm_spi_init();
 }
 
 /*
@@ -510,8 +537,6 @@
 
 	if (!tpm_is_supported())
 		return 0;
-
-	tpm_spi_init();
 
 	ret = tpm_tis_wait_init(locality);
 	if (ret)
diff --git a/include/boot_rom.h b/include/boot_rom.h
index d63f31f..817c740 100644
--- a/include/boot_rom.h
+++ b/include/boot_rom.h
@@ -134,5 +134,8 @@
 void boot_rom_bringup_ppboot(void);
 void boot_rom_init_boot_info(void);
 int boot_rom_verify_image(enum hl_fw_boot_ptr_imgs img_type);
+void boot_rom_tpm_spi_global_ctx_init(struct spi_ctx *tpm_spi);
+void boot_rom_tpm_spi_init(void);
+void boot_rom_tpm_spi_setup(void);
 
 #endif /* _BOOT_ROM_H_ */
